Source/WebCore/platform/mediarecorder/cocoa/MediaRecorderPrivateWriterWebM.mm

```diff
@@ -0,0 +1,601 @@
+                    protectedThis->partiallyFlushEncodedQueues();
+                }
+            });
+        });
+
+        m_resumedVideoTime = MonotonicTime::now();
+
+        maybeStartWriting();
+    }
+
+    if (!m_hasStartedWriting) // We don't record video frames received before the first audio frame.
+        return;
+
+    assertIsCurrent(MainThreadDispatcher::singleton());
+
+    // We take the time before m_firstVideoFrameAudioTime is set so that the first frame will always apepar to have a timestamp of 0 but with a longer duration.
+    auto sampleTime = m_firstVideoFrameAudioTime ? nextVideoFrameTime() : resumeVideoTime();
+
+    if (!m_firstVideoFrameAudioTime)
+        m_firstVideoFrameAudioTime = sampleTime;
+
+    ASSERT(m_lastEnqueuedRawVideoFrame <= sampleTime);
+    ASSERT(m_lastEnqueuedRawVideoFrame >= m_lastReceivedCompressedVideoFrame);
+    m_lastEnqueuedRawVideoFrame = sampleTime;
+    m_pendingVideoFrames.append({ Ref { frame }, sampleTime });
+    LOG(MediaStream, "appendVideoFrame:enqueuing raw video frame:%f queue:%zu first:%f last:%f", sampleTime.value(), m_pendingVideoFrames.size(), m_pendingVideoFrames.first().second.value(), m_pendingVideoFrames.last().second.value());
+
+    encodePendingVideoFrames();
+}
+
+Seconds MediaRecorderPrivateWriterWebM::nextVideoFrameTime() const
+{
+    ASSERT(!m_hasAudio || m_firstVideoFrameAudioTime);
+    auto now = MonotonicTime::now();
+    auto frameTime = now - m_resumedVideoTime;
+    LOG(MediaStream, "nextVideoFrameTime: frameTime:%f nextVideoFrameTime:%f", frameTime.value(), (frameTime + *m_firstVideoFrameAudioTime).value());
+    frameTime = frameTime + *m_firstVideoFrameAudioTime;
+    // Take the max value from m_lastAudioSampleTime to handle the occasional wall clock time drift and ensure muxed frames times
+    // is always monotonically increasing.
+    return std::max(frameTime, Seconds { m_lastAudioSampleTime.toDouble() });
+}
+
+Seconds MediaRecorderPrivateWriterWebM::resumeVideoTime() const
+{
+    return m_hasAudio ? Seconds { m_resumedAudioTime.toDouble() } : m_currentVideoDuration;
+}
+
+Seconds MediaRecorderPrivateWriterWebM::sampleTime(const RetainPtr<CMSampleBufferRef>& sample) const
+{
+    return Seconds { PAL::CMTimeGetSeconds(PAL::CMSampleBufferGetPresentationTimeStamp(sample.get())) };
+}
+
+void MediaRecorderPrivateWriterWebM::encodePendingVideoFrames()
+{
+    assertIsCurrent(MainThreadDispatcher::singleton());
+
+    if (m_pendingVideoFrames.isEmpty())
+        return;
+
+    m_pendingVideoEncode++;
+
+    m_currentOperation = Ref { m_currentOperation }->whenSettled(MainThreadDispatcher::singleton(), [weakThis = ThreadSafeWeakPtr { *this }, this, pendingVideoFrame = std::exchange(m_pendingVideoFrames, { })]() mutable {
```

Comment:
I would tend to only set m_currentOperation for flush operations.
Flush promise ensures that all encodes are done.

Something like:
if (m_currentOperation) {
   m_currentOperation->whenSettled(... { encodePendingVideoFrames(); })
   return;
}
while (! m_pendingVideoFrames.isEmpty())
   m_videoEncoder->encode(...);

Reply:
we want to encode the frames in the right order. So if new frames are coming in (while a flush is pending) we would output the current in-flight frame (they were the one available at the time of flush).

So we need to only start encoding the new frames once the previous ones have finished being encoded.

If we didn't append them to the current operation we would have frames out of order at the end.

Reply:
One more thing: I use the ability to know if there's an encode operation mid-flight or not to determine if it's okay to immediately write the audio frames to the container. 
Using the promise only for when flushing would lose that. 
Unless I called flush regularly. but those block the main thread for audio, so don't want to flush too often 

(in an earlier version I wasn't flushing the audio compressor when fetchData was called for this reason) I will take this as a follow up, making AudioSampleBufferCompressor::flush return a promise instead.

Reply:
Ok, I rewrote it , removing serialisation of decoding operations, I guess they could all happen in parallel and up to the encoder to serialise them as it already does.

It does rely however on the VideoEncoder to allow to call encode() before the previous promise resolved. I don't think it may always be the case.